src/imitation/algorithms/bc.py

```diff
@@ -198,85 +183,69 @@ def train(
           on_epoch_end: optional callback to run at
             the end of each epoch. Will receive all locals from this function as
             dictionary argument (!!).
+          log_interval: log stats after every log_interval batches
         """
         assert self.batch_size >= 1
+        total_samples = n_epochs * self.expert_dataset.size()
+        total_batches = np.ceil(total_samples / self.batch_size)
         samples_so_far = 0
         batch_num = 0
         for epoch_num in trange(n_epochs, desc="BC epoch"):
             while samples_so_far < (epoch_num + 1) * self.expert_dataset.size():
+                BaseAlgorithm._update_current_progress_remaining(
+                    self, batch_num, total_batches
+                )
+                BaseAlgorithm._update_learning_rate(self, self.policy.optimizer)
                 batch_num += 1
                 batch_dict = self.expert_dataset.sample(self.batch_size)
                 assert len(batch_dict["obs"]) == self.batch_size
                 samples_so_far += self.batch_size
-                feed_dict = {
-                    self._true_acts_ph: batch_dict["acts"],
-                    self.policy.obs_ph: batch_dict["obs"],
-                }
-                _, stats_dict = self.sess.run(
-                    [self._train_op, self._stats_dict], feed_dict=feed_dict
-                )
+
+                obs_tensor = th.as_tensor(batch_dict["obs"]).to(self.policy.device)
```

Comment:
IMO this kind of conversion (e.g. to one-hot) belongs in the algorithm. You might want different kinds of preprocessing for different algorithms, depending on for example what kind of observation/action spaces they can support. Whereas things like frame stacking, normalization belong outside the algorithm (perhaps in dataset wrappers).

Reply:
The preprocessing @qxcv mentioned already [happens inside `Policy` automatically](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L82-L91) when we call `policy.evaluate_actions(obs, acts)`. (Image observations are scaled, discrete observations are transformed to one-hot, etc.).

Reply:
I suspect that we don't need to convert to Tensor here either, as there's some code I saw in the policy implementation which converts observation to Tensor for us. https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L237

Reply:
@qxcv : The Policy will automatically work with images if we set `policy_cls=CnnPolicy` in `BC.__init__`.

Reply:
@shwang I initially shared your intuition that it wouldn't be necessary to convert to a Tensor here because some other part of the code would do it automatically. But I got an error when I didn't convert, so I added the conversion to fix the error.